autogpt/llm/base.py

```diff
@@ -73,17 +74,37 @@ class ChatSequence:
     """Utility container for a chat sequence"""
 
     model: ChatModelInfo
-    messages: list[Message] = field(default_factory=list)
+    messages: list[Message] = field(default_factory=list[Message])
 
-    def __getitem__(self, i: int):
-        return self.messages[i]
+    @overload
+    def __getitem__(self, key: int) -> Message:
+        ...
+
+    @overload
+    def __getitem__(self, key: slice) -> ChatSequence:
+        ...
+
+    def __getitem__(self, key: int | slice):
+        if isinstance(key, slice):
+            copy = deepcopy(self)
+            copy.messages = self.messages[key]
```

Comment:
I think you want 
```suggestion
            return ChatSequence(
                model=self.model,
                messages=deepcopy(self.messages[key]),
            )
```
so you're not copying the entire message history just to take a slice.

Reply:
Then it should be 
```suggestion
            return self.__class__(
                model=self.model,
                messages=deepcopy(self.messages[key]),
                **kwargs,
            )
```
Because otherwise, `message_history[1:10]` will return a `ChatSequence`. Where to get the `**kwargs` from though? Subclasses may (and do) have more initializing attributes that need to be included. This is not an issue by just doing `deepcopy(self)`.

Reply:
Fair, that's a thorny problem.  Leave as is, then, I think.